### ISSUE TITLE
Can't install on OSX

### ISSUE THREAD

**Issue by stubelshe** [nfarina/homebridge#638]
I've seen this issue posted here, followed a number of instructions and am not making progress.

Running: `sudo npm install -g homebridge`

Gets me:

```
curve25519@1.1.0 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/curve25519
node-gyp rebuild

shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
sh: node-gyp: command not found

bignum@0.11.0 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/srp/node_modules/bignum
node-gyp configure build

shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
sh: node-gyp: command not found

mdns@2.3.3 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/mdns
node-gyp rebuild

shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
sh: node-gyp: command not found

ed25519@0.0.3 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/ed25519
node-gyp rebuild

shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
sh: node-gyp: command not found
npm ERR! Darwin 15.4.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "-g" "homebridge"
npm ERR! node v4.4.4
npm ERR! npm  v2.5.1
npm ERR! code ELIFECYCLE

npm ERR! curve25519@1.1.0 install: `node-gyp rebuild`
npm ERR! Exit status 127
npm ERR! 
npm ERR! Failed at the curve25519@1.1.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the curve25519 package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls curve25519
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/stubelshe/npm-debug.log
```

As I said, I've followed a number of instruction on these boards and others and had no progress, any ideas?

Thanks

**Comment by jer78** [comment]
Have you installed XCode and Homebrew and Node.js first? 

Here's some instructions to get you going. 

1. Install XCode
2. Install HomeBrew
    From terminal type: ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
3. Install Node.js 
    From terminal type: brew install node
    To verify the versions type: node -v npm -v
4. Install HomeBridge
    Type: sudo npm install -g homebridge

**Comment by stubelshe** [comment]
Updates:

Xcode v.7.3.1 i already installed

Homebrew was already installed, to ensure I ran your command;
`It appears Homebrew is already installed. If your intent is to reinstall you
should do the following before running this installer again:
    ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/uninstall)"
The current contents of /usr/local are bin Cellar CODEOFCONDUCT.md CONTRIBUTING.md etc git include lib Library LICENSE.txt mysql-5.5.21-osx10.6-x86_64 mysql-5.6.10-osx10.7-x86 opt README.md share SUPPORTERS.md .git .gitignore`

Node was already installed, to endure I ran `node -v`;
`v4.4.4`

Npm was also installed; I ran `npm -v`;
`2.5.1`

I then ran `sudo npm install -g homebridge` and got this;
`> curve25519@1.1.0 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/curve25519
> node-gyp rebuild

shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
sh: node-gyp: command not found

> bignum@0.11.0 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/srp/node_modules/bignum
> node-gyp configure build

shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
sh: node-gyp: command not found

> ed25519@0.0.3 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/ed25519
> node-gyp rebuild

shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
sh: node-gyp: command not found

> mdns@2.3.3 install /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/node_modules/mdns
> node-gyp rebuild

shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
sh: node-gyp: command not found
npm ERR! Darwin 15.4.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "-g" "homebridge"
npm ERR! node v4.4.4
npm ERR! npm  v2.5.1
npm ERR! code ELIFECYCLE

npm ERR! curve25519@1.1.0 install: `node-gyp rebuild`
npm ERR! Exit status 127
npm ERR! 
npm ERR! Failed at the curve25519@1.1.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the curve25519 package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls curve25519
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/stubelshe/npm-debug.log`

**Comment by jer78** [comment]
Sorry for the late response. It looks to me like there's permissions issues. Are you logging in as an Administrator account on your Mac Mini? When you run the sudo command, are you getting prompted for your password?

**Comment by apwelsh** [comment]
Don't use homebrew to install nodejs.  Go to HTTP://nodejs.org and download the pkg, and install it. 

Then use the sudo npm -g homebridge command.

It will throw some security warnings that root does not have access to .node-gyp folders and then a follow-up that it is attempting reinstall using temp dev dir.  Then it should work.

Make sure you have write access to /use/local/* folders.  Because of rootless mode in OS X El Capitan, the root user no longer gets full access to everything.

Depending on your version of brew, it may be trying to do work in /usr instead of /usr/local

**Comment by tannebil** [comment]
Yes and yes.

**Comment by tannebil** [comment]
That's the process I successfully followed.

**Comment by stubelshe** [comment]
@jer78 yes, I mad working as an admin and it does ask me for a password. It's actually a Mad Pro if it matters, haha.

**Comment by stubelshe** [comment]
Here's my latest status;

Went to nodejs.org and downloaded the latest
Ran `node -v`;
`v6.2.0`

Ran `npm -v`;
`3.8.9`

I then ran `sudo npm install -g homebridge` and got this;

```
/usr/local/bin/homebridge -> /usr/local/lib/node_modules/homebridge/bin/homebridge

mdns@2.3.3 install /usr/local/lib/node_modules/homebridge/node_modules/mdns
node-gyp rebuild

shell-init: error retrieving current directory: getcwd: cannot access parent directories: Permission denied
sh: node-gyp: command not found
/usr/local/lib
└── (empty)

npm ERR! Darwin 15.4.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "-g" "homebridge"
npm ERR! node v6.2.0
npm ERR! npm  v3.8.9
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn

npm ERR! mdns@2.3.3 install: `node-gyp rebuild`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the mdns@2.3.3 install script 'node-gyp rebuild'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the mdns package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs mdns
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls mdns
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/stubelshe/npm-debug.log
npm ERR! code 1
```

**Comment by apwelsh** [comment]
I used an old iMac I have that has home-brew, but not node.js installed.  I used home-brew to install node.js.  `sudo brew install nodejs`

The result was it installed a very old version:
```
Warning: You are using OS X 10.11.
We do not provide support for this pre-release version.
You may encounter build failures or other breakage.
==> Downloading https://homebrew.bintray.com/bottles/node-0.12.7_1.el_capitan.bottle.tar.gz
######################################################################## 100.0%
==> Pouring node-0.12.7_1.el_capitan.bottle.tar.gz
==> Caveats
Bash completion has been installed to:
  /usr/local/etc/bash_completion.d
==> Summary
🍺  /usr/local/Cellar/node/0.12.7_1: 2780 files, 30M
```

Then I used npm to install homebridge.  `sudo npm -g install homebridge` which installed just fine.  It did generate warning, because root is not longer god-mode user account due to rootless mode of OS X, but it recovers by using a temporary directory to build the package and deploy it.  

Now that I know it works, I removed homebridge `sudo npm -g uninstall homebridge` and node.js `sudo brew remove nodejs` to undo my testing.

Next I installed node.js using the pkg from the website, because it's just so nice to have this on my systems. 

So, knowing it's not homebrew, node.js, nor homebridge, what can be causing your problem?  Most likely answer is rootless.  But don't turn it off, instead, just make sure to setup your directory permissions correctly.

In your output your first error reports that /usr/local/lib is empty.  Rather than provide a bunch of testing, lets just run the commands to create the structure and assign ownerships correctly...


```
sudo mkdir -p /usr/local/bin /usr/local/lib /usr/local/share /usr/local/etc
sudo chown root:admin /usr/local
sudo chown root:admin /usr/local/bin /usr/local/lib /usr/local/share /usr/local/etc
sudo find /usr/local -type d -exec chmod 755
```

this should fix the important directories, and create any missing ones.  if you had home-brew installed before upgrading to El Capitan, then you had it running in a mode that required root.  The updated version, when installed on a clean install of El Capitan support rootless mode, and permissions the /usr/local folder appropriately.  Anyways, give this a try, and if you cannot get in, we should look at your directory permissions much closer.

**Comment by ncosper** [comment]
I know this thread is about a mo th old but I ran into the exact same problems. Maybe this will help someone in the future. 
The problem that I found was that my hard drive had a space In its name. This caused issues installing. I removed the space and it installed instantly. Big potential problems when you do this with home folders and such but that fixes it. I'm not sure if there is a fix using commands to install but that was my experience.
